examples/transformer/model.py

```diff
@@ -0,0 +1,195 @@
+            vocab_size=self.vocab_size,
+            output_layer=self.word_embedder.embedding,
+            hparams=self.config_model.decoder,
+        )
+
+        self.smoothed_loss_func = LabelSmoothingLoss(
+            label_confidence=self.config_model.loss_label_confidence,
+            tgt_vocab_size=self.vocab_size,
+            ignore_index=0,
+        )
+
+    def forward(  # type: ignore
+        self,
+        encoder_input: torch.Tensor,
+        is_train_mode: Optional[bool],
+        decoder_input: Optional[torch.LongTensor] = None,
+        labels: Optional[torch.LongTensor] = None,
+        beam_width: Optional[int] = None,
+    ):
+        r"""TODO: Add detailed docstrings.
+        """
+
+        batch_size = encoder_input.size(0)
+        # (text sequence length excluding padding)
+        encoder_input_length = (encoder_input != 0).int().sum(dim=1)
+
+        if is_train_mode:
+            self.train()
+        else:
+            self.eval()
+
+        # Source word embedding
+        src_word_embeds = self.word_embedder(encoder_input)
+        src_word_embeds = src_word_embeds * self.config_model.hidden_dim ** 0.5
+
+        # Position embedding (shared b/w source and target)
+        src_seq_len = torch.full(
+            (batch_size,), encoder_input.size(1), dtype=torch.int32
+        )
+        src_seq_len = src_seq_len.to(device=encoder_input.device)
+
+        src_pos_embeds = self.pos_embedder(sequence_length=src_seq_len)
+        src_input_embedding = src_word_embeds + src_pos_embeds
+
+        encoder_output = self.encoder(
+            inputs=src_input_embedding, sequence_length=encoder_input_length
+        )
+
+        if is_train_mode:
+            assert decoder_input is not None
+            assert labels is not None
+
+            tgt_word_embeds = self.word_embedder(decoder_input)
+            tgt_word_embeds = (
+                tgt_word_embeds * self.config_model.hidden_dim ** 0.5
+            )
+            tgt_seq_len = decoder_input.new_full(
+                (batch_size,), decoder_input.size(1),
+            )
+
+            tgt_pos_embeds = self.pos_embedder(sequence_length=tgt_seq_len)
+
+            tgt_input_embedding = tgt_word_embeds + tgt_pos_embeds
+
+            # For training
+            outputs = self.decoder(
+                memory=encoder_output,
+                memory_sequence_length=encoder_input_length,
+                inputs=tgt_input_embedding,
+                decoding_strategy="train_greedy",
+            )
+            label_lengths = (labels != 0).long().sum(dim=1)
+            is_target = (labels != 0).float()
+            mle_loss = self.smoothed_loss_func(
+                outputs.logits, labels, label_lengths
+            )
+            mle_loss = (mle_loss * is_target).sum() / is_target.sum()
+            return mle_loss
+
+        else:
+            start_tokens = encoder_input.new_full(
+                (batch_size,), self.vocab.bos_token_id,
+            )
```

Comment:
Other examples like `seq2seq_attn` also create `start_tokens`, but using different APIs. This one looks concise and should be adopted in other examples.

Reply:
The `seq2seq_attn` creates `start_tokens` by:
https://github.com/asyml/texar-pytorch/blob/ec3be3824f959ee0a919d754e65923e973fd00ae/examples/seq2seq_attn/seq2seq_attn.py#L101-L102
I don't really see the difference. And I also think the above should be changed to using `torch.full_like` instead of creates ones then multiplying.

Reply:
@gpengzhi can you update `seq2seq_attn` like this?